### DISCUSSION TITLE
Can't register new user - 419 page expired error.

### DISCUSSION THREAD

**Discussion by tomlawesome** [firefly-iii/firefly-iii#4699] [Firefly III]
Re-opening as it turns out the issue is not resolved at all:

If I login via HTTP headers, I can see and load all the pages, but I cannot setup anything. It errors to a 419 Page Expired error. The same thing happens if I try to use the web authentication, and try to register an initial admin account. I believe this is related to CORS/CSP somehow, but my CSP is non-existent, and I have tried opening my CORS settings as wide as I can go. 

I am running Firefly III latest docker version

**Description**
Using Eloquent authentication, fresh install, brand new Postgres database (deleted + new database created twice, to no change)

I can see that there are no tables, I then boot the docker image, zero errors occur, once booted I can browse via my asubdomain on my FQDN to my firefly install no problem and checking Postgres for new tables in the dedicated Firefly database shows it has populated the database with tables etc. On filling out the initial user registration (for admin account per docs) I get a 419 'Page Expired' error.

**Extra info**
Postgres latest version, Firefly latest version, docker + docker compose latest versions.

**Bonus points**

    I have read the FAQ at https://bit.ly/FF3-FAQ
    Screenshot: https://i.ibb.co/m4xDbNn/firefly.png
    Log file from "docker logs firefly" https://pb.tomlawson.io/?1f3d921aa4f0566e#8uARESUDSD35i2QwW9cFMvyqcZ17XN7KXNoMSbUkCPGh
    I was able to replicate the issue on the demo site: No, because of the nature of the problem.

**Comment by tomlawesome** [comment]
Since the original post, I've made FireflyIII available locally on a port. This is not a CORS/Headers issue, as the 419 occurs direct to the server IP/port. 

I have tried:

- Changing the 'upload' storage folder permissions (didn't think it would have an effect, but trying anything now)
- Various CORS/headers settings, which is now a moot point
- Deleting the container and re-pulling the image
- Changing the entire docker container entrypoint directory (/var/www/html) to chmod 755 to check for perms issues. (have since recreated the container). 

Questions:

- Does the app require a specific PUID/PGUID for the docker container? 
- Does FireflyIII make use of CSRF tokens? 
- What user has permissions to clear the php server cache? I get the following error when trying to do it as root:
```
root@021135ec734d:/var/www/html# php artisan cache:clear
Failed to clear cache. Make sure you have the appropriate permissions.
```
- Does FireflyIII forward the X-CSRF-TOKEN header or expect it or something? I think this is the root cause of my problem, but given that it happens on a local IP/port I can't understand how I've affected anything?

**Reply by JC5** [reply]
Yes, firefly III uses CSRF tokens. You can inspect the source over at the demo site and see how the token responds to a page refresh.

**Comment by tomlawesome** [comment]
Two weeks on and still have the same issue with no resolution, guess I just can't use firefly III which is a shame as it looks great. This is the only error I can find anywhere, but unfortunately it does not give any indication as to the route cause that I can tell?

![image](https://user-images.githubusercontent.com/76453276/117052370-8efd1f00-ad0f-11eb-9101-5cee62d6f972.png)

**Reply by JC5** [reply]
Sorry, I completely forgot about this ticket. I see you're behind cloudflare but that shouldn't matter much. Allow me to dive into this tomorrow.

**Reply by tomlawesome** [reply]
> Sorry, I completely forgot about this ticket. I see you're behind cloudflare but that shouldn't matter much. Allow me to dive into this tomorrow.

No worries! Thanks, I do really appreciate it! Cloudflare shouldn’t be an issue as I still get the 419 on a local ip/port direct connection. 

It’s the docker image running on a minimal Debian10 VM, launched by compose. Will add my compose when I’m back at the pc!

**Reply by JC5** [reply]
Is there anything like a reverse proxy in front of the image?

**Reply by tomlawesome** [reply]
No, it happens even when it’s just docker in a VM with the port exposed and connecting directly via LAN

**Reply by JC5** [reply]
What are the cookie environment variables? Everything that starts with cookie_ basically.

**Reply by tomlawesome** [reply]
> What are the cookie environment variables? Everything that starts with cookie_ basically.

Sorry for the slow reply -- had an enjoyable evening with a RAID card issue. 
```
COOKIE_PATH="/"
COOKIE_DOMAIN=
COOKIE_SECURE=false
COOKIE_SAMESITE=lax
```

**Reply by JC5** [reply]
Alright that looks good at least. What are the mappings?

Just a minute ago I launched and have running a new Firefly III instance. There were my commands:

```
curl https://raw.githubusercontent.com/firefly-iii/docker/main/docker-compose.yml > docker-compose.yml
curl https://raw.githubusercontent.com/firefly-iii/firefly-iii/main/.env.example > .env
docker-compose -f docker-compose.yml up -d
```

In the source of localhost/register (the page you end up with) the csrf-token is static. It doesn't change. The nonce does, but the token doesn't:

```
    <meta name="csrf-token" content="s08Gxn72QCES4scFcPUV6JnK3trf4QydUS7j3AlG">
        <link href="v1/lib/bs/css/bootstrap.min.css?v=5.5.10" rel="stylesheet" type="text/css" nonce="kQ71CVDId+cPQmwls05OrA==">
```

How is this with your instance?

**Reply by tomlawesome** [reply]
Morning. I've removed authelia protection, so the instance should be available at https://fly.tomlawon.io now. 

```
<meta name="csrf-token" content="80kA0CssJ6PvNXDQzUMa9dUCCFA35rwfWfUqWvWA">
<meta content='width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no' name='viewport'>
<link href="v1/lib/bs/css/bootstrap.min.css?v=5.5.8" rel="stylesheet" type="text/css" nonce="6oh2QFuJC5HYy7OvqS0kwg==">
<link href="v1/lib/fa/css/font-awesome.min.css?v=5.5.8" rel="stylesheet" type="text/css" nonce="6oh2QFuJC5HYy7OvqS0kwg==">
<link href="v1/css/daterangepicker.css?v=5.5.8" rel="stylesheet" type="text/css" nonce="6oh2QFuJC5HYy7OvqS0kwg==">
<link href="v1/lib/adminlte/css/AdminLTE.min.css?v=5.5.8" rel="stylesheet" type="text/css" nonce="6oh2QFuJC5HYy7OvqS0kwg==">
<link href="v1/lib/adminlte/css/skins/skin-blue-light.min.css?v=5.5.8" rel="stylesheet" type="text/css" nonce="6oh2QFuJC5HYy7OvqS0kwg==">
<link href="v1/css/firefly.css?v=5.5.8" rel="stylesheet" type="text/css" nonce="6oh2QFuJC5HYy7OvqS0kwg==">
```

**Reply by tomlawesome** [reply]
```
curl https://raw.githubusercontent.com/firefly-iii/docker/main/docker-compose.yml > docker-compose.yml
curl https://raw.githubusercontent.com/firefly-iii/firefly-iii/main/.env.example > .env
docker-compose -f docker-compose.yml up -d
```

This runs perfectly on a local IP/port. So it must be within my config somewhere. The most notable element is that I'm using (or trying to use) postgres rather than MySQL. I've got to head out, but I'll try switching it to a postgres DB later and see if the issue reappears, and also try switching my main install to MySQL to rule it out.

**Reply by JC5** [reply]
Thanks. The site doesn't work yet but no matter. Postgres shouldn't matter (but you never know). Do you map any folders in the Docker container?

**Reply by tomlawesome** [reply]
> 
> 
> Thanks. The site doesn't work yet but no matter. Postgres shouldn't matter (but you never know). Do you map any folders in the Docker container?

Sorry should have said fly.tomlawson.io is on my main server and I've not changed any setup so it will still return 419 error on registering. Test instance @ https://flytest.tomlawson.io/

```
  firefly:
    image: jc5x/firefly-iii:latest
    env_file: $DOCKERDIR/appdata/firefly/.env.firefly
    container_name: firefly
    restart: always
    security_opt:
 - no-new-privileges:true
    ports:
 - "8128:8080"
    profiles:
 - isolation
    networks:
 - Mars
    volumes:
 - $DOCKER_APPDATA/firefly/data:/var/www/html/storage/upload
 - $DOCKERDIR/certs/firefly/server.crt:/certs/server.crt:ro
 - $DOCKERDIR/certs/firefly/server.key:/certs/server.key:ro
 - $DOCKERDIR/certs/firefly/serverCA.crl:/certs/serverCA.crl:ro
 - $DOCKERDIR/certs/firefly/serverCA.crt:/certs/serverCA.crt:ro
 - $DOCKERDIR/certs/synoldap/synology_ldap.crt:/certs/synology_ldap:ro
 - $DOCKERDIR/certs/synoldap/synology_ldap.key:/certs/synology_ldap.key:ro
 - $DOCKERDIR/certs/synoldap/syno_ldap_CA.crt:/certs/syno_ldap_CA.crt:ro
```

Usually I use traeifk labels, but have exposed a port to access it locally to negate any reverse proxy issues. I think the issue is in my .env config?

.env file:
```
# DATABASE
DB_CONNECTION=pgsql
DB_HOST=postgres
DB_PORT=5432
DB_DATABASE=firefly
DB_USERNAME=firefly
DB_PASSWORD=<REDACTED>
PGSQL_SSL_MODE=require                                                                  # PostgreSQL supports SSL. You can configure it here.
PGSQL_SSL_ROOT_CERT=/certs/serverCA.crt                                                 # If you use Docker or similar, you can set these variables from a file by appending them with _FILE
PGSQL_SSL_CERT=/certs/server.crt
PGSQL_SSL_KEY=/certs/server.key
PGSQL_SSL_CRL_FILE=/certs/serverCA.crl

# If you're looking for performance improvements, you could install memcached.
CACHE_DRIVER=memcached
SESSION_DRIVER=memcached
# can be tcp, unix or http
REDIS_SCHEME=tcp
REDIS_HOST=redis
REDIS_PORT=6379
REDIS_PASSWORD=<REDACTED>                                                 # always use quotes and make sure redis db "0" and "1" exists. Otherwise change accordingly.
REDIS_DB="0"
REDIS_CACHE_DB="1"

COOKIE_PATH="/"
COOKIE_DOMAIN=
COOKIE_SECURE=false
COOKIE_SAMESITE=lax

# SMTP
MAIL_MAILER=log
MAIL_HOST=${SMTP_HOST}
MAIL_PORT=${SMTP_PORT}
MAIL_FROM=sysadmin.firefly@tomlawson.io
MAIL_USERNAME=${SMTP_USERNAME}
MAIL_PASSWORD=${SMTP_NOREPLY_PASS}
MAIL_ENCRYPTION=${SMTP_TLS}
SEND_REGISTRATION_MAIL=true
SEND_ERROR_MESSAGE=true
SEND_LOGIN_NEW_IP_WARNING=true
SEND_REPORT_JOURNALS=true                                                                # These messages contain (sensitive) transaction information:

# GEOLOCATION
MAPBOX_API_KEY=<REDACTED>
ENABLE_EXTERNAL_MAP=false
MAP_DEFAULT_LAT=51.983333                                                                 # The map will default to this location:
MAP_DEFAULT_LONG=5.916667
MAP_DEFAULT_ZOOM=6

LOGIN_PROVIDER=eloquent
AUTHENTICATION_GUARD=web
AUTHENTICATION_GUARD_HEADER=REMOTE_USER
AUTHENTICATION_GUARD_EMAIL=
DISABLE_FRAME_HEADER=false
DISABLE_CSP_HEADER=false

TRACKER_SITE_ID=
TRACKER_URL=
SEND_TELEMETRY=false
ALLOW_WEBHOOKS=false
DKR_CHECK_SQLITE=true
DKR_RUN_MIGRATION=true
DKR_RUN_UPGRADE=true
DKR_RUN_VERIFY=true
DKR_RUN_REPORT=true
DKR_RUN_PASSPORT_INSTALL=true

APP_NAME=firefly
ADLDAP_CONNECTION=default
BROADCAST_DRIVER=log
QUEUE_DRIVER=sync
CACHE_PREFIX=firefly
PUSHER_KEY=
IPINFO_TOKEN=
PUSHER_SECRET=
PUSHER_ID=
DEMO_USERNAME=
DEMO_PASSWORD=
IS_HEROKU=false
FIREFLY_III_LAYOUT=v1
APP_URL=http://fly.tomlawson.io 
```

**Reply by tomlawesome** [reply]
![image](https://user-images.githubusercontent.com/76453276/117275121-a0e8da00-ae55-11eb-9dc0-931e5c09eab1.png)


Database does seem fine. Requests show up etc.

**Reply by JC5** [reply]
I'm not sure if this is a C/P error but make sure that the comments are on new lines, not at the end of the lines.

Or better, remove all comments from the .env file. Just in case.

Also this:

```
CACHE_DRIVER=memcached
SESSION_DRIVER=memcached
```
You also refer to redis later on. Is memcached installed and working? What if you use `file`, the default value?

**Reply by tomlawesome** [reply]
So, in the interests of just getting it running I copied the stock .env over and only filled in some basic info like app_key and database details. I also switched to MySQL to rule out a Postgres issue, not that I could see it being that and am still getting 419? 

The database is populated with tables etc and reports no errors in MySQL:
![image](https://user-images.githubusercontent.com/76453276/117317155-2a61d180-ae81-11eb-8dd4-87fc631d67a1.png)

```
APP_ENV=local
APP_DEBUG=false
SITE_OWNER=mail@example.com
APP_KEY=
DEFAULT_LANGUAGE=en_GB
DEFAULT_LOCALE=equal
TZ=Europe/London
TRUSTED_PROXIES=**
LOG_CHANNEL=stack
APP_LOG_LEVEL=notice
AUDIT_LOG_LEVEL=info

DB_CONNECTION=mysql
DB_HOST=jupiterdb
DB_PORT=3306
DB_DATABASE=firefly
DB_USERNAME=firefly
DB_PASSWORD=

MYSQL_USE_SSL=false
MYSQL_SSL_VERIFY_SERVER_CERT=true
MYSQL_SSL_CAPATH=/etc/ssl/certs/
MYSQL_SSL_CA=
MYSQL_SSL_CERT=
MYSQL_SSL_KEY=
MYSQL_SSL_CIPHER=

#PGSQL_SSL_MODE=require
#PGSQL_SSL_ROOT_CERT=/certs/serverCA.crt
#PGSQL_SSL_CERT=/certs/server.crt
#PGSQL_SSL_KEY=/certs/server.key
#PGSQL_SSL_CRL_FILE=/certs/serverCA.crl

CACHE_DRIVER=file
SESSION_DRIVER=file

#REDIS_SCHEME=tcp
#REDIS_PATH=
#REDIS_HOST=127.0.0.1
#REDIS_PORT=6379
#REDIS_PASSWORD=null
#REDIS_DB="0"
#REDIS_CACHE_DB="1"

COOKIE_PATH="/"
COOKIE_DOMAIN=
COOKIE_SECURE=false
COOKIE_SAMESITE=lax

#MAIL_MAILER=log
#MAIL_HOST=null
#MAIL_PORT=2525
#MAIL_FROM=changeme@example.com
#MAIL_USERNAME=null
#MAIL_PASSWORD=null
#MAIL_ENCRYPTION=null

#SEND_REGISTRATION_MAIL=true
#SEND_ERROR_MESSAGE=true
#SEND_LOGIN_NEW_IP_WARNING=true
#SEND_REPORT_JOURNALS=true

#MAPBOX_API_KEY=
#ENABLE_EXTERNAL_MAP=false
#MAP_DEFAULT_LAT=51.983333
#MAP_DEFAULT_LONG=5.916667
#MAP_DEFAULT_ZOOM=6
#
LOGIN_PROVIDER=eloquent
AUTHENTICATION_GUARD=web
AUTHENTICATION_GUARD_HEADER=REMOTE_USER
AUTHENTICATION_GUARD_EMAIL=

CUSTOM_LOGOUT_URI=

#ADLDAP_CONNECTION_SCHEME=OpenLDAP
#ADLDAP_AUTO_CONNECT=true
#ADLDAP_CONTROLLERS=
#ADLDAP_PORT=389
#ADLDAP_TIMEOUT=5
#ADLDAP_BASEDN=""
#ADLDAP_FOLLOW_REFFERALS=false

#ADLDAP_USE_SSL=false
#ADLDAP_USE_TLS=false
#ADLDAP_SSL_CACERTDIR=
#ADLDAP_SSL_CACERTFILE=
#ADLDAP_SSL_CERTFILE=
#ADLDAP_SSL_KEYFILE=
#ADLDAP_SSL_CIPHER_SUITE=
#ADLDAP_SSL_REQUIRE_CERT=
#ADLDAP_ADMIN_USERNAME=
#ADLDAP_ADMIN_PASSWORD=
#ADLDAP_ACCOUNT_PREFIX=
#ADLDAP_ACCOUNT_SUFFIX=
#ADLDAP_PASSWORD_SYNC=false
#ADLDAP_LOGIN_FALLBACK=false
#ADLDAP_DISCOVER_FIELD=distinguishedname
#ADLDAP_AUTH_FIELD=distinguishedname
#ADLDAP_SYNC_FIELD=userprincipalname

DISABLE_FRAME_HEADER=false
DISABLE_CSP_HEADER=false
TRACKER_SITE_ID=
TRACKER_URL=
SEND_TELEMETRY=false
ALLOW_WEBHOOKS=false

DKR_CHECK_SQLITE=true
DKR_RUN_MIGRATION=true
DKR_RUN_UPGRADE=true
DKR_RUN_VERIFY=true
DKR_RUN_REPORT=true
DKR_RUN_PASSPORT_INSTALL=true

APP_NAME=FireflyIII
ADLDAP_CONNECTION=default
BROADCAST_DRIVER=log
QUEUE_DRIVER=sync
CACHE_PREFIX=firefly
PUSHER_KEY=
IPINFO_TOKEN=
PUSHER_SECRET=
PUSHER_ID=
DEMO_USERNAME=
DEMO_PASSWORD=
IS_HEROKU=false
FIREFLY_III_LAYOUT=v1
APP_URL=http://localhost
```

**Reply by JC5** [reply]
Especially with the file driver I'm running out of ideas rapidly. What happens when you enable dev mode on cloudflare? I have my own install also behind cloudflare but using vanilla nginx.

**Reply by tomlawesome** [reply]
> Especially with the file driver I'm running out of ideas rapidly. What happens when you enable dev mode on cloudflare? I have my own install also behind cloudflare but using vanilla nginx.

Success! 

Some good rubber-ducking later, and I realised there is a CSP header variable in the .env file:

```
DISABLE_CSP_HEADER=false
```

With the above set to 'true', FireflyIII loads and runs perfectly, no more 419 error. Using Postgres, behind traefik etc etc. I really think this was it all along!

![image](https://user-images.githubusercontent.com/76453276/117355591-b6d4ba00-aeaa-11eb-9366-84ad33c56519.png)

**Reply by JC5** [reply]
That is so weird. That header does nothing related to this issue. I thought. Oh well, glad to see it finally works 😁

**Reply by IAmTomahawkx** [reply]
I know this is a 4 year old thread but I also just encountered this issue, and the resolution here worked for me as well. Docker install with cloudflare + Caddy in front. Was able to register and set up the app, however I restarted the app to enable webhooks and suddenly started receiving 419 when submitting login. Not sure exactly why this fixes it, but fix it it did. Thanks!
